File: day-3/Q539.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>

using namespace std;

class Solution
{
public:
    int findMinDifference(vector<string>& timePoints) {
        vector<int> timeValues(timePoints.size());
        for (int i = 0; i < timePoints.size(); ++i) {
            int colonIdx = timePoints[i].find(':');
            int hours = stoi(timePoints[i].substr(0, colonIdx));
            int minutes = stoi(timePoints[i].substr(colonIdx + 1));

            minutes += hours * 60;
            timeValues[i] = minutes;
        }
        sort(timeValues.begin(), timeValues.end());
        
        int minDiff = INT_MAX;
        for (int i = 1; i < timeValues.size(); ++i) {
            minDiff = min(minDiff, timeValues[i] - timeValues[i - 1]);
        }

        return min(minDiff, 1440 - timeValues.back() + timeValues[0]);
    }
};

int main() {
    Solution sol;
    vector<string> timePoints;
    timePoints.push_back("00:00");
    timePoints.push_back("23:59");
    timePoints.push_back("00:00");
    cout << sol.findMinDifference(timePoints) << endl;
    return -1;
}

File: day-3/Q539.java
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays; 

class Solution {
    public int findMinDifference(List<String> timePoints) {
        int[] minutes = new int[timePoints.size()];
        for (int i = 0; i < timePoints.size(); i++) {
            String time = timePoints.get(i);
            int h = Integer.parseInt(time.substring(0, 2));
            int m = Integer.parseInt(time.substring(3));
            minutes[i] = h * 60 + m;
        }

        Arrays.sort(minutes);

        int ans = Integer.MAX_VALUE;
        for (int i = 0; i < minutes.length - 1; i++) {
            ans = Math.min(ans, minutes[i + 1] - minutes[i]);
        }

        return Math.min(
            ans,
            24 * 60 - minutes[minutes.length - 1] + minutes[0]
        );
    }
}

class Q539 {
    public static void main(String[] args) {
        Solution sol = new Solution();
        List<String> timePoints = new ArrayList<>();
        timePoints.add("23:59");
        timePoints.add("05:00");
        timePoints.add("22:59");
        System.out.println(sol.findMinDifference(timePoints));
    }
}